scripts/scons_helpers/build-3rdparty.py

```diff
@@ -664,18 +664,152 @@ def splitenv(envlist):
+    'Cygwin'
+    >>> openssl_get_platform('')
+    'cc'
+    """
+
+    # we assume that:
+    # - 'android' has a higher priority than 'linux'
+    # - if toolchain contains '64' and 'android' => we have android64 on 64-bit CPU
+    # - if toolchain contains '64' and 'linux' => we have linux64 on 64-bit CPU
+    # - openssl architectures 'aout', 'elf', 'latomic' are meaningless (they cannot be found in
+    #   toolchain)
+    # - compiler suffixes '-cc', '-gcc', '-clang' are meaningless too
+    # - aix* OSes are never used :)
+    # - 'android64' are currently just aliases for 'android' architectures;
+    #   'android64-aarch64' is the same as 'android-arm64'
+
+    bitness = 64 if '64' in toolchain else 32
+
+    ### corner cases
+
+    # if android toolchain contains both 'arm' and 'eabi' => 'arm' (32 bit)
+    # ('armeabi' doesn't work)
+    if 'android' in toolchain and 'arm' in toolchain and 'eabi' in toolchain:
+        return 'android-arm'
+    # 32 bit ARM => linux-generic32; 'linux-armv4' fails :(
+    if 'linux' in toolchain and 'arm' in toolchain and bitness == 32:
+        return 'linux-generic32'
+
+    ### generic detection
+
+    # XXX: this is a dirty parser based on very simple heuristics and verified on a
+    # limited set of platforms
+
+    # Order is significant here as we search substrings in a whole toolchain string.
+    #
+    # `arch` is an architecture that openssl accepts for a given `os`. arch in a toolchain may
+    # be the same but called differently, e.g. 'x86' vs 'i686'. To handle such cases,
+    # `arch_aliases` are used (see below): if `arch` is not in toolchain, we try to search all
+    # aliases (one by one) before moving to the next arch.
+    platforms = [
+        # [(os,          [arch, ...]), ...]
+        ('BSD',          ['aarch64', 'generic64', 'generic32', 'ia64', 'riscv64',
+                          'sparc64', 'sparcv8', 'x86_64', 'x86']),
+        ('Cygwin',       ['i386', 'i486', 'i586', 'i686', 'x86_64', 'x86']),
+        #('android64',    ['aarch64', 'mips64', 'x86_64']),
+        ('android',      ['arm64', 'armeabi', 'arm', 'mips64', 'mips', 'x86_64', 'x86']),
+        ('darwin64',     ['arm64', 'ppc', 'x86_64']),
+        ('darwin',       ['i386', 'ppc']),
+        ('linux64',      ['loongarch64', 'mips64', 'riscv64', 's390x', 'sparcv9']),
+        ('linux32',      ['s390x']),
+        ('linux',        ['aarch64', 'arm64ilp32', 'armv4', 'generic64', 'generic32', 'ia64',
+                          'mips64', 'mips32', 'ppc64le', 'ppc64', 'ppc', 'sparcv8', 'sparcv9',
+                          'x86_64', 'x86', 'x32']),
+        ('mingw64',      []),
+        ('mingw',        []),
+    ]
+    arch_aliases = {
+        'armv4':     ['arm32', 'arm'],
+        'generic32': ['generic'],
+        'mips32':    ['mips'],
+        'x86':       ['i386', 'i486', 'i586', 'i686'],
+        'arm64':     ['aarch64'], # fix for absent platform dirs in android NDK
+    }
```

Comment:
1. IIRC, we should never pass `linux-armv4` to OpenSSL, right? Because compilation will fail. Instead, we should use `linux-generic32`, which we handle above. In this case, I guess we should **remove `armv4`** from `platforms` and `arch_aliases`.

2. Since in OpenSSL, `android-armeabi` is an alias for `android-arm` (you mentioned it in chat), then I suggest to **remove `armeabi`** from `android` platform.

3. Also maybe **remove** commented out **`android64`** block? Since `android64` has no real value.

4. I think we should **remove `generic64` and `generic32`** from `BSD` and `linux`, and from `arch_aliases`. They will never occur in toolchain string.

Reply:
I suggest to use `generic` as a fallback for BSD and linux OSes.

For example, if we detected that `os` is `linux`, and can't detect `arch`, then we should use `linux-generic32` or `linux-generic64` depending on bitness.

This way, this special check:

```
    if 'linux' in toolchain and 'arm' in toolchain and bitness == 32:
        return 'linux-generic32'
```

will be naturally not needed, because we will automatically employ `linux-generic32`.

Reply:
It seems that this special check:

```
    if 'android' in toolchain and 'arm' in toolchain and 'eabi' in toolchain:
        return 'android-arm'
```

is not needed now. If `android` and `arm` are in toolchain, we will use `android-arm`, regardless of `eabi`, which is correct.

Reply:
Yes, there are problems with armv4. On Android, only the following 4 archs are acceptable (see component after `arch-` string): `arm`, `arm64`, `x86`, and `x86_64`:
````
# docker run --rm -ti --cap-add SYS_PTRACE -u 0 -v .../roc/roc-toolkit:/opt/roc -w /opt/roc -e CI= rocstreaming/toolchain-linux-android:ndk21 /bin/bash
root@eb758900a4b5:/opt/roc# ls -l /opt/android-ndk-r21/platforms/android-29/     
total 16
drwxr-xr-x 3 root root 4096 Jan  6  2020 arch-arm
drwxr-xr-x 3 root root 4096 Jan  6  2020 arch-arm64
drwxr-xr-x 3 root root 4096 Jan  6  2020 arch-x86
drwxr-xr-x 3 root root 4096 Jan  6  2020 arch-x86_64
root@eb758900a4b5:/opt/roc# 
```
Removed `armv4` and others.

Reply:
Special check
```
if 'android' in toolchain and 'arm' in toolchain and 'eabi' in toolchain:
```
is still required.

Reply:
> I suggest to use generic as a fallback for BSD and linux OSes.

done